src/memtable.rs

```diff
@@ -1189,30 +1281,69 @@ impl Default for MemTableRecoverContext<VacantAllocator> {
 
 impl<A: AllocatorTrait> ReplayMachine for MemTableRecoverContext<A> {
     fn replay(&mut self, mut item_batch: LogItemBatch, file_id: FileId) -> Result<()> {
-        for item in item_batch.iter() {
-            match &item.content {
-                LogItemContent::Command(Command::Clean)
-                | LogItemContent::Command(Command::Compact { .. }) => {
-                    self.log_batch.push((*item).clone());
-                }
-                LogItemContent::Kv(KeyValue { op_type, .. }) if *op_type == OpType::Del => {
-                    self.log_batch.push((*item).clone());
-                }
-                _ => {}
+        if file_id.queue == LogQueue::Append {
+            let mut new_tombstones = Vec::new();
+            self.memtables
+                .replay_append_writes(item_batch.drain().filter(|item| {
+                    if Self::is_tombstone(item) {
+                        new_tombstones.push(item.clone());
+                    }
+                    true
+                }));
+            self.tombstone_items.append(&mut new_tombstones);
+        } else {
+            let mut new_tombstones = Vec::new();
+            let mut is_group = None;
+            let items = item_batch
+                .drain()
+                .filter(|item| {
+                    if let Some(g) = AtomicGroupStatus::parse(item) {
+                        if is_group.is_none() {
+                            is_group = Some(g);
+                        } else {
+                            debug_assert!(false, "skipped an atomic group: {:?}", g);
+                        }
+                        return false;
+                    }
+                    if Self::is_tombstone(item) {
+                        new_tombstones.push(item.clone());
+                    }
+                    true
+                })
+                .collect();
+            if let Some((id, status)) = is_group {
+                self.accept_new_group(
+                    file_id.queue,
+                    id,
+                    PendingAtomicGroup {
+                        status,
+                        items,
+                        tombstone_items: new_tombstones,
+                    },
+                );
+            } else {
+                self.tombstone_items.append(&mut new_tombstones);
+                self.memtables.replay_rewrite_writes(items.into_iter());
             }
         }
-        match file_id.queue {
-            LogQueue::Append => self.memtables.replay_append_writes(item_batch.drain()),
-            LogQueue::Rewrite => self.memtables.replay_rewrite_writes(item_batch.drain()),
-        }
         Ok(())
     }
 
     fn merge(&mut self, mut rhs: Self, queue: LogQueue) -> Result<()> {
-        self.log_batch.merge(&mut rhs.log_batch.clone());
+        self.tombstone_items
+            .append(&mut rhs.tombstone_items.clone());
+        for (id, groups) in rhs.pending_atomic_groups.drain() {
```

Comment:
Is it possible the merge is not done continuously? For example, (Begin, Middle), (Middle, Middle), (Middle, End) are merged as (Middle, Middle), (Middle, End), (Begin, Middle).

Reply:
No, it's guaranteed to merge neighbors.